truss/tests/test_control_truss_patching.py

```diff
@@ -0,0 +1,310 @@
+
+    def predict_with_python_requirement_removed(req):
+        th.remove_python_requirement(req)
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_system_requirement_added(pkg):
+        th.add_system_package(pkg)
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_system_requirement_removed(pkg):
+        th.remove_system_package(pkg)
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    def predict_with_ignored_changes():
+        top_pycache_path = custom_model_control / "__pycache__"
+        top_pycache_path.mkdir()
+        (top_pycache_path / "bla.pyc").touch()
+        model_pycache_path = custom_model_control / "model" / "__pycache__"
+        model_pycache_path.mkdir()
+        (model_pycache_path / "foo.pyc").touch()
+        return th.docker_predict([1], tag=tag, binary=binary)
+
+    with ensure_kill_all():
+        result = th.docker_predict([1], tag=tag, binary=binary)
+        assert result[0] == 1
+        orig_num_truss_images = len(th.get_all_docker_images())
+
+        result = predict_with_updated_model_code()
+        assert result[0] == 2
+        assert orig_num_truss_images == current_num_docker_images(th)
+
+        result = predict_with_added_empty_directory()
+        assert result[0] == 2
+        assert orig_num_truss_images == current_num_docker_images(th)
+
+        container = th.get_running_serving_container_ignore_hash()
+
+        python_req = "pydot"
+        result = predict_with_python_requirement_added(python_req)
+        assert result[0] == 2
+        assert current_num_docker_images(th) == orig_num_truss_images
+        verify_python_requirement_installed_on_container(container, python_req)
+
+        result = predict_with_python_requirement_removed(python_req)
+        assert result[0] == 2
+        assert current_num_docker_images(th) == orig_num_truss_images
+        verify_python_requirement_not_installed_on_container(container, python_req)
+
+        system_pkg = "jq"
+        result = predict_with_system_requirement_added(system_pkg)
+        assert result[0] == 2
+        assert current_num_docker_images(th) == orig_num_truss_images
+        verify_system_package_installed_on_container(container, system_pkg)
+
+        result = predict_with_system_requirement_removed(system_pkg)
+        assert result[0] == 2
+        assert current_num_docker_images(th) == orig_num_truss_images
+        verify_system_requirement_not_installed_on_container(container, system_pkg)
+
+        result = predict_with_unpatchable_change()
+        assert result[0] == 2
+        assert current_num_docker_images(th) == orig_num_truss_images + 1
+
+        result = predict_with_ignored_changes()
+        assert result[0] == 2
+        assert current_num_docker_images(th) == orig_num_truss_images + 1
+
+
+@pytest.mark.skip(reason="Unsupported patch")
+@pytest.mark.integration
```

Comment:
do all of these need to be integration tests? Can we just make the test suite be unit tests that check if the right patch was created or if there is an error.

These integration tests might be worth keeping, but unit tests would bring down the iteration cycling and regression detection time to almost free.

Reply:
Yeah the idea was to add in the unit tests when I get around to implementing the patching support for each of these – I'll keep these around to test the end to end flow